File: python/numbers/unit_converter.py
"""Unit Converter — convert between measurement units."""


# Base units: meters (distance), grams (weight), liters (volume).
# Temperature is special-cased with formulas.

# Conversion factors TO base unit.
_TO_BASE = {
    # distance -> meters
    "km": 1000.0,
    "m": 1.0,
    "mi": 1609.344,
    "ft": 0.3048,
    # weight -> grams
    "kg": 1000.0,
    "g": 1.0,
    "lb": 453.592,
    "oz": 28.3495,
    # volume -> liters
    "L": 1.0,
    "gal": 3.78541,
    "mL": 0.001,
    "cup": 0.236588,
}

_CATEGORIES = {
    "distance": {"km", "m", "mi", "ft"},
    "weight": {"kg", "g", "lb", "oz"},
    "volume": {"L", "gal", "mL", "cup"},
    "temperature": {"celsius", "fahrenheit", "kelvin"},
}

_UNIT_TO_CATEGORY = {}
for cat, units in _CATEGORIES.items():
    for u in units:
        _UNIT_TO_CATEGORY[u] = cat


def _to_celsius(value: float, unit: str) -> float:
    if unit == "celsius":
        return value
    if unit == "fahrenheit":
        return (value - 32.0) * 5.0 / 9.0
    if unit == "kelvin":
        return value - 273.15
    raise ValueError(f"unknown temperature unit: {unit}")


def _from_celsius(value: float, unit: str) -> float:
    if unit == "celsius":
        return value
    if unit == "fahrenheit":
        return value * 9.0 / 5.0 + 32.0
    if unit == "kelvin":
        return value + 273.15
    raise ValueError(f"unknown temperature unit: {unit}")


def convert(value: float, from_unit: str, to_unit: str) -> float:
    """Convert a value from one unit to another.

    Supports temperature (celsius, fahrenheit, kelvin),
    distance (km, mi, m, ft), weight (kg, lb, oz, g),
    and volume (L, gal, mL, cup).
    """
    if from_unit == to_unit:
        return value

    from_cat = _UNIT_TO_CATEGORY.get(from_unit)
    to_cat = _UNIT_TO_CATEGORY.get(to_unit)

    if from_cat is None:
        raise ValueError(f"unknown unit: {from_unit}")
    if to_cat is None:
        raise ValueError(f"unknown unit: {to_unit}")
    if from_cat != to_cat:
        raise ValueError(f"cannot convert {from_unit} ({from_cat}) to {to_unit} ({to_cat})")

    if from_cat == "temperature":
        return _from_celsius(_to_celsius(value, from_unit), to_unit)

    # Ratio-based: convert to base, then from base to target.
    base_value = value * _TO_BASE[from_unit]
    return base_value / _TO_BASE[to_unit]


if __name__ == "__main__":
    demos = [
        (100, "celsius", "fahrenheit"),
        (1, "mi", "km"),
        (2.5, "kg", "lb"),
        (1, "gal", "L"),
    ]
    for val, fr, to in demos:
        result = convert(val, fr, to)
        print(f"{val} {fr} = {result:.4f} {to}")

File: rust/numbers/unit_converter.rs
//! Unit Converter — convert between measurement units.

/// Return the category and base-unit factor for a ratio-based unit.
/// Base units: meters (distance), grams (weight), liters (volume).
fn unit_info(unit: &str) -> Option<(&'static str, f64)> {
    match unit {
        // distance -> meters
        "km" => Some(("distance", 1000.0)),
        "m"  => Some(("distance", 1.0)),
        "mi" => Some(("distance", 1609.344)),
        "ft" => Some(("distance", 0.3048)),
        // weight -> grams
        "kg" => Some(("weight", 1000.0)),
        "g"  => Some(("weight", 1.0)),
        "lb" => Some(("weight", 453.592)),
        "oz" => Some(("weight", 28.3495)),
        // volume -> liters
        "L"  => Some(("volume", 1.0)),
        "gal" => Some(("volume", 3.78541)),
        "mL" => Some(("volume", 0.001)),
        "cup" => Some(("volume", 0.236588)),
        _ => None,
    }
}

/// Convert to celsius from the given temperature unit.
fn to_celsius(value: f64, unit: &str) -> Result<f64, String> {
    match unit {
        "celsius"    => Ok(value),
        "fahrenheit" => Ok((value - 32.0) * 5.0 / 9.0),
        "kelvin"     => Ok(value - 273.15),
        _ => Err(format!("unknown temperature unit: {unit}")),
    }
}

/// Convert from celsius to the given temperature unit.
fn from_celsius(value: f64, unit: &str) -> Result<f64, String> {
    match unit {
        "celsius"    => Ok(value),
        "fahrenheit" => Ok(value * 9.0 / 5.0 + 32.0),
        "kelvin"     => Ok(value + 273.15),
        _ => Err(format!("unknown temperature unit: {unit}")),
    }
}

fn is_temperature(unit: &str) -> bool {
    matches!(unit, "celsius" | "fahrenheit" | "kelvin")
}

/// Convert a value from one unit to another.
///
/// Supports temperature (celsius, fahrenheit, kelvin),
/// distance (km, mi, m, ft), weight (kg, lb, oz, g),
/// and volume (L, gal, mL, cup).
pub fn convert(value: f64, from: &str, to: &str) -> Result<f64, String> {
    if from == to {
        return Ok(value);
    }

    // Temperature: formula-based.
    if is_temperature(from) || is_temperature(to) {
        if !is_temperature(from) || !is_temperature(to) {
            return Err(format!("cannot convert {from} to {to}"));
        }
        return from_celsius(to_celsius(value, from)?, to);
    }

    // Ratio-based: normalize to base unit, then to target.
    let (from_cat, from_factor) = unit_info(from)
        .ok_or_else(|| format!("unknown unit: {from}"))?;
    let (to_cat, to_factor) = unit_info(to)
        .ok_or_else(|| format!("unknown unit: {to}"))?;

    if from_cat != to_cat {
        return Err(format!("cannot convert {from} ({from_cat}) to {to} ({to_cat})"));
    }

    Ok(value * from_factor / to_factor)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn celsius_to_fahrenheit() {
        let r = convert(100.0, "celsius", "fahrenheit").unwrap();
        assert!((r - 212.0).abs() < 1e-10);
    }

    #[test]
    fn fahrenheit_to_celsius() {
        let r = convert(32.0, "fahrenheit", "celsius").unwrap();
        assert!(r.abs() < 1e-10);
    }

    #[test]
    fn celsius_to_kelvin() {
        let r = convert(0.0, "celsius", "kelvin").unwrap();
        assert!((r - 273.15).abs() < 1e-10);
    }

    #[test]
    fn miles_to_km() {
        let r = convert(1.0, "mi", "km").unwrap();
        assert!((r - 1.609344).abs() < 1e-4);
    }

    #[test]
    fn kg_to_lb() {
        let r = convert(1.0, "kg", "lb").unwrap();
        assert!((r - 2.20462).abs() < 0.001);
    }

    #[test]
    fn gallon_to_liters() {
        let r = convert(1.0, "gal", "L").unwrap();
        assert!((r - 3.78541).abs() < 0.001);
    }

    #[test]
    fn same_unit() {
        assert_eq!(convert(42.0, "m", "m").unwrap(), 42.0);
    }

    #[test]
    fn cross_category_error() {
        assert!(convert(1.0, "km", "kg").is_err());
    }

    #[test]
    fn unknown_unit_error() {
        assert!(convert(1.0, "parsec", "km").is_err());
    }
}
